canbench-rs/src/lib.rs

```diff
@@ -596,22 +597,131 @@ pub fn bench_scope(name: &'static str) -> BenchScope {
     BenchScope::new(name)
 }
 
+/// Benchmarks the scope this function is declared in.
+///
+/// NOTE: It's important to assign this function, otherwise benchmarking won't work correctly.
+///
+/// # Correct Usage
+///
+/// ```
+/// # #[repr(u16)]
+/// # enum Scope {
+/// #     MyScope = 0,
+/// # }
+///
+/// fn my_func() {
+///   let _p = canbench_rs::bench_scope_id(Scope::MyScope as u16);
+///   // Do something.
+/// }
+/// ```
+///
+/// # Incorrect Usages
+///
+/// ```
+/// # #[repr(u16)]
+/// # enum Scope {
+/// #     MyScope = 0,
+/// # }
+///
+/// fn my_func() {
+///   let _ = canbench_rs::bench_scope_id(Scope::MyScope as u16); // Doesn't capture the scope.
+///   // Do something.
+/// }
+/// ```
+///
+/// ```
+/// # #[repr(u16)]
+/// # enum Scope {
+/// #     MyScope = 0,
+/// # }
+///
+/// fn my_func() {
+///   canbench_rs::bench_scope_id(Scope::MyScope as u16); // Doesn't capture the scope.
+///   // Do something.
+/// }
+/// ```
+#[must_use]
+pub fn bench_scope_id(code: u16) -> BenchScope {
+    BenchScope::from_code(code)
+}
+
+pub trait ScopeIdName {
+    /// Returns the name of the scope given its ID.
+    fn name_from_id(id: u16) -> Option<&'static str>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum ScopeId {
+    Name(&'static str),
+    Code(u16),
+}
+
+impl ScopeId {
+    fn from_text(text: &'static str) -> Self {
+        Self::Name(text)
+    }
+
+    fn from_code(code: u16) -> Self {
+        Self::Code(code)
+    }
+}
+
+/// Sets the resolver for the scope ID to name mapping.
+///
+/// This function should be called before any benchmarks are run.
+pub fn set_bench_id_resolver<T: ScopeIdName + 'static>() {
```

Comment:
I'm not quite sure how this is supposed to be used. Is it the canbench binary or the user that would have to set this resolver? Should there be a default so that it's easy to use (as an end user I probably don't want to have to think about this resolver).